src/index.js

```diff
@@ -1,174 +1,231 @@
+    } else {
+      // socket has no rooms set, create and add the room
+      sids.set(id, new Set([room]))
+    }
+    // if room has a set of sockets, just add the new socket id
+    if (socketsInRoom instanceof Set) {
+      socketsInRoom.add(id);
+    } else {
+      // room has no set of sockets, create and add it
+      rooms.set(room, new Set([id]));
+    }
+    // add to wild rooms
+    // TODO visit wild logic to work Map and Set
+    // wild.add(room);
+    return Promise.resolve({success: true});
   }
 
   /**
    * Remove a socket from a room or from all rooms if a room is not passed.
-   *
-   * @param {String} id
-   * @param {String} room
-   * @param {Function} [fn]
+   * Used for leaving the rooms
+   * @param {String} id - spark's ID
+   * @param {String} room - the room to leave
    * @api public
    */
 
-  function del(id, room, fn = noop) {
-    let prune = pruner(id)
+  function del(id, room) {
+    // if (room) {
+    //   // TODO update logic for wildcard
+    //   if (wildcard && ~room.indexOf('*')) {
+    //     let rms = keys(id ? sids[id] || {} : rooms)
+    //     wild.find(room, rms, prune)
+    //   } else if (sids[id] && sids[id][room]) {
+    //     prune(room)
+    //   }
+    // } else {
+    //   for (room in sids[id]) prune(room)
+    // }
+    // setImmediate(fn, null, null)
+    
     if (room) {
-      if (wildcard && ~room.indexOf('*')) {
-        let rms = keys(id ? sids[id] || {} : rooms)
-        wild.find(room, rms, prune)
-      } else if (sids[id] && sids[id][room]) {
-        prune(room)
-      }
-    } else {
-      for (room in sids[id]) prune(room)
+      return prune(id, room);
     }
-    setImmediate(fn, null, null)
+    let promises = [];
+    rooms.forEach((value, key) => {
+      promises.push(prune(id, key));
+    });
+    return Promise.all(promises);
   }
 
   /**
    * Get client ids connected to this room.
-   *
-   * @param {String} room
-   * @param {Function} fn
+   * Used to get clients in a room
+   * @param {String} room - the room to get the clients connected to
+   * @param {Function} fn - callback
    * @api public
    */
 
-  function clients(room, fn = noop) {
-    room = rooms[room] || {}
-    setImmediate(fn, null, keys(room))
+  function clients(room) {
+    let socketSet = rooms.get(room);
+    if (socketSet instanceof Set) {
+      return Promise.resolve(Array.from(socketSet));
+    }
+    return Promise.resolve([]);
   }
 
   /**
    * Remove all sockets from a room.
-   *
-   * @param {String|Array} room
-   * @param {Function} [fn]
+   * Used to empty the room
+   * @param {String|String[]} room - the room(s) to empty
+   * @param {Function} [fn] - callback
    * @api public
    */
 
-  function empty(room, fn = noop) {
-    if (room && rooms[room]) {
-      if (!isArray(room)) clear(room)
-      else for (const rm of room) clear(rm)
+  function empty(room) {
+    if (isArray(room)) {
+      let promises = [];
+      for (let rm in rooms) promises.push(clear(rm));
```

Comment:
I think there is a typo. If I'm not wrong `rooms` should be `room`. Also how about using `Array.prototype.map()`?